ext/filter/logical_filters.c

```diff
@@ -796,35 +796,70 @@ static int _php_filter_validate_ipv6(char *str, size_t str_len) /* {{{ */
 				return 0;
 			}
 			if (*str == ':') {
-				if (compressed) {
+				if (compressed_pos >= 0) {
 					return 0;
 				}
-				blocks++; /* :: means 1 or more 16-bit 0 blocks */
-				compressed = 1;
-
+				if (ip && blocks < 8) {
+					ip[blocks] = -1;
+				}
+				compressed_pos = blocks++; /* :: means 1 or more 16-bit 0 blocks */
 				if (++str == end) {
-					return (blocks <= 8);
+					if (blocks > 8) {
+						return 0;
+					} else {
+						goto fixup_ip;
+					}
 				}
 			} else if ((str - 1) == s) {
 				/* don't allow leading : without another : following */
 				return 0;
 			}
 		}
-		n = 0;
+		num = n = 0;
 		while ((str < end) &&
 		       ((*str >= '0' && *str <= '9') ||
 		        (*str >= 'a' && *str <= 'f') ||
 		        (*str >= 'A' && *str <= 'F'))) {
+			if (ip) {
```

Comment:
I'd compute this unconditionally and add an `else { break; }` so we don't repeat these conditions.

Reply:
Ah, that's nice! Thanks.